src/main/groovy/com/github/jengelman/gradle/plugins/shadow/tasks/ShadowJar.java

```diff
@@ -100,6 +103,11 @@ public ShadowJar dependencies(Action<DependencyFilter> c) {
         return this;
     }
 
+    public ShadowJar entryAction(Action<ZipEntry> action) {
```

Comment:
We're gonna need to make our own 1st class citizen for `ZipEntry`. We can't expose the apache classes as part of the public API for shadow because those classes are relocated during packaging.

But overall this PR looks good. I'll likely merge it this way and then figure out the stuff above.